SpringKonstant/src/org/usfirst/frc6357/robotcode/subsystems/TFMini.java

```diff
@@ -204,25 +170,65 @@ private void runLiDAR()
             {
                while(startRead)
                {
-            	   if(amountOfBytes() >= 9)
-            	   {
-            		   byte bytes[] = readBytes();
-            		   if(isValid(bytes))
-            		   {
-            			   distL = bytes[2];
-            			   distH = bytes[3];
-            			   strengthL = bytes[4];
-            			   strengthH = bytes[5];
-            			   reservedBytes = bytes[6];
-            			   signalQualityDegree = bytes[7];
-            		   }
-            	   }
+                   byte bytes[] = sensor.read(1);
+                   if(bytes.length > 0)
+                   {
+                       switch(state)
+                       {
+                           case WAIT_START1:
+                               if(bytes[0] == 0x59)
+                               {
+                                   count = 0;
+                                   checksum = bytes[0];
+                                   packet[0] = bytes[0];
+                                   state = ReceivingState.WAIT_START2;
+                               }
+                               break;
+                           case WAIT_START2:
+                               if(bytes[0] == 0x59)
+                               {
+                                   count++;
+                                   checksum += bytes[0];
+                                   packet[1] = bytes[0];
+                                   state = ReceivingState.RECEIVING;
+                               }
+                               else
+                               {
+                                   state = ReceivingState.WAIT_START1;
+                               }
+                               break;
+                           case RECEIVING:
+                               count++;
+                               packet[count] = bytes[0];
+                               if(count == PACKET_LENGTH - 1)
+                               {
+                                   // Reached the end of the packet, check checksum
+                                   if(checksum == bytes[0])
+                                   {
+                                       parsePacket();
+                                   }
+                                   state = ReceivingState.WAIT_START1;
+                               }
+                               else
+                               {
+                                   checksum += bytes[0];
+                               }
+                               break;
+                       }
+                   }
                }
             }
             catch(Exception e) {}
         }).start();
     }
     
+    public void parsePacket()
+    {
+        distance = (((int) packet[3] * 256) + ((int) packet[2]));
```

Comment:
I think these should be uints, else 0xFF = -128 rather than the 256 we want it to be.

Reply:
You are correct (except for the 0xFF = -128 thing coz it actually equals -1 if we're using 8 bit signed integers or 255 for unsigned.) My assumption, though, is that "byte" is unsigned so the 0xFF we read there will be cast to 255 as an integer (signed or unsigned). Maybe Java works differently from C, of course?